File: packages/sdk-python/netcrawl/daemon/spawner.py
"""
netcrawl/daemon/spawner.py

Spawns and manages unit subprocesses.
"""

import os
import sys
import json
import subprocess


_active_processes: dict[str, subprocess.Popen] = {}


def spawn_unit(
    unit_id: str,
    script_path: str,
    class_name: str,
    api_url: str,
    injected_fields: dict,
) -> int:
    """
    Spawn a Python unit subprocess.
    Returns the PID.
    """
    env = os.environ.copy()
    env.update({
        "NETCRAWL_UNIT_ID": unit_id,
        "NETCRAWL_API_URL": api_url,
        "NETCRAWL_SCRIPT_PATH": script_path,
        "NETCRAWL_CLASS_NAME": class_name,
        "NETCRAWL_INJECTED": json.dumps(injected_fields),
    })

    process = subprocess.Popen(
        [sys.executable, "-m", "netcrawl.runner"],
        env=env,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
    )

    _active_processes[unit_id] = process
    print(f"[spawner] Spawned unit {unit_id} (PID {process.pid})")
    return process.pid


def kill_unit(unit_id: str) -> bool:
    """Terminate a running unit process."""
    proc = _active_processes.get(unit_id)
    if proc is None:
        return False

    proc.terminate()
    try:
        proc.wait(timeout=5)
    except subprocess.TimeoutExpired:
        proc.kill()

    del _active_processes[unit_id]
    print(f"[spawner] Killed unit {unit_id}")
    return True


def get_unit_status(unit_id: str) -> str:
    """Returns 'running', 'stopped', 'crashed', or 'unknown'."""
    proc = _active_processes.get(unit_id)
    if proc is None:
        return "unknown"

    poll = proc.poll()
    if poll is None:
        return "running"
    elif poll == 0:
        return "stopped"
    else:
        return "crashed"


def list_active() -> list[dict]:
    return [
        {"unit_id": uid, "pid": proc.pid, "status": get_unit_status(uid)}
        for uid, proc in _active_processes.items()
    ]

File: packages/sdk-js/src/daemon/spawner.ts
/**
 * daemon/spawner.ts
 *
 * Spawns and manages worker subprocesses using child_process.fork.
 */

import { fork, type ChildProcess } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

interface ActiveProcess {
  process: ChildProcess;
  workerId: string;
}

const _activeProcesses: Map<string, ActiveProcess> = new Map();

/**
 * Spawn a Node.js worker subprocess.
 * Returns the PID.
 */
export function spawnWorker(
  workerId: string,
  scriptPath: string,
  className: string,
  apiUrl: string,
  injectedFields: Record<string, unknown>,
): number {
  const runnerPath = path.resolve(__dirname, '..', 'runner.js');

  const env: Record<string, string> = {
    ...process.env as Record<string, string>,
    NETCRAWL_WORKER_ID: workerId,
    NETCRAWL_API_URL: apiUrl,
    NETCRAWL_SCRIPT_PATH: scriptPath,
    NETCRAWL_CLASS_NAME: className,
    NETCRAWL_INJECTED: JSON.stringify(injectedFields),
  };

  const child = fork(runnerPath, [], {
    env,
    stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
  });

  const pid = child.pid ?? -1;

  _activeProcesses.set(workerId, { process: child, workerId });

  child.on('exit', () => {
    _activeProcesses.delete(workerId);
  });

  if (child.stdout) {
    child.stdout.on('data', (data: Buffer) => {
      process.stdout.write(data);
    });
  }
  if (child.stderr) {
    child.stderr.on('data', (data: Buffer) => {
      process.stderr.write(data);
    });
  }

  console.log(`[spawner] Spawned worker ${workerId} (PID ${pid})`);
  return pid;
}

/**
 * Terminate a running worker process.
 */
export function killWorker(workerId: string): boolean {
  const entry = _activeProcesses.get(workerId);
  if (!entry) {
    return false;
  }

  entry.process.kill('SIGTERM');

  // Force kill after 5 seconds
  const killTimer = setTimeout(() => {
    try {
      entry.process.kill('SIGKILL');
    } catch {
      // Already dead
    }
  }, 5000);

  entry.process.on('exit', () => {
    clearTimeout(killTimer);
  });

  _activeProcesses.delete(workerId);
  console.log(`[spawner] Killed worker ${workerId}`);
  return true;
}

/**
 * Returns 'running', 'stopped', 'crashed', or 'unknown'.
 */
export function getWorkerStatus(workerId: string): string {
  const entry = _activeProcesses.get(workerId);
  if (!entry) {
    return 'unknown';
  }

  if (entry.process.exitCode === null && !entry.process.killed) {
    return 'running';
  } else if (entry.process.exitCode === 0) {
    return 'stopped';
  } else {
    return 'crashed';
  }
}

export function listActive(): Array<{ workerId: string; pid: number; status: string }> {
  return Array.from(_activeProcesses.entries()).map(([wid, entry]) => ({
    workerId: wid,
    pid: entry.process.pid ?? -1,
    status: getWorkerStatus(wid),
  }));
}
